src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs

```diff
@@ -42,11 +46,41 @@ namespace NetTopologySuite.Operation.Union
     /// <seealso href="http://code.google.com/p/nettopologysuite/issues/detail?id=44"/>
     public class CascadedPolygonUnion
     {
-        private ICollection<Geometry> _inputPolys;
-        private GeometryFactory _geomFactory;
-#if UseWorker
-        private int _numThreadsStarted = 0;
-#endif
+        /// <summary>
+        /// A union strategy that uses the classic NTS {@link SnapIfNeededOverlayOp},
+        /// and for polygonal geometries a robustness fallback using <c>Buffer(0)</c>.
+        /// </summary>
+        internal static readonly UnionStrategy ClassicUnion = new UnionStrategy(
+            (g0, g1) =>
+            {
+                try
+                {
+                    return SnapIfNeededOverlayOp.Union(g0, g1);
+                }
+                catch (TopologyException ex)
+                {
+                    // union-by-buffer only works for polygons
+                    if (g0.Dimension != Dimension.Surface || g1.Dimension != Dimension.Surface)
+                        throw ex;
```

Comment:
instead of this conditional rethrow (which should be `throw;` instead of `throw ex;`), use a filtered catch:

```csharp
catch (TopologyException) when (g0.Dimension == Dimension.Surface && g1.Dimension == Dimension.Surface)
{
    return UnionPolygonsByBuffer(g0, g1);
}
```